plugins/l4lb/l4lb.go

```diff
@@ -0,0 +1,281 @@
+// Copyright 2015 CNI authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net"
+	"os"
+	"runtime"
+
+	"github.com/dcos/dcos-cni/pkg/spartan"
+
+	"github.com/vishvananda/netlink"
+
+	"github.com/containernetworking/cni/pkg/invoke"
+	"github.com/containernetworking/cni/pkg/ip"
+	"github.com/containernetworking/cni/pkg/ipam"
+	"github.com/containernetworking/cni/pkg/ns"
+	"github.com/containernetworking/cni/pkg/skel"
+	"github.com/containernetworking/cni/pkg/types"
+	"github.com/containernetworking/cni/pkg/types/current"
+	"github.com/containernetworking/cni/pkg/version"
+)
+
+type NetConf struct {
+	types.NetConf
+	Args     map[string]interface{} `json:"args"`
+	MTU      int                    `json:"mtu"`
+	Delegate map[string]interface{} `json:"delegate"`
+}
+
+func init() {
+	// This ensures that main runs only on main thread (thread group leader).
+	// Since namespace ops (unshare, setns) are done for a single thread, we
+	// must ensure that the goroutine does not jump from OS thread to thread
+	runtime.LockOSThread()
+}
+
+func setupContainerVeth(netns, ifName string, mtu int, pr *current.Result, spartanIPs []net.IPNet) (string, error) {
+	// The IPAM result will be something like IP=192.168.3.5/24,
+	// GW=192.168.3.1. What we want is really a point-to-point link but
+	// veth does not support IFF_POINTOPONT. So we set the veth
+	// interface to 192.168.3.5/32. Since the device netmask is set to
+	// /32, this would not add any routes to the main routing table.
+	// Therefore, in order to reach the spartan interfaces, we will have
+	// to explicitly set routes to the spartan interface through this
+	// device.
+
+	var hostVethName string
+
+	err := ns.WithNetNSPath(netns, func(hostNS ns.NetNS) error {
```

Comment:
Again CNI idioms. I actually agree with you on this. This does make the code a lot more cryptic. Let me define the closure outside the function call.